### PR TITLE
feature/os2loop analytics

### PR DESCRIPTION
**Pull request by martinyde**
https://jira.itkdev.dk/browse/LOOP-752
- LOOP-752: Added analytics modul
- LOOP-752: Added helper service and attachment function

### CODE REVIEW

File: web/profiles/custom/os2loop/modules/os2loop_analytics/src/Helper/AnalyticsHelper.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    foreach ($scripts as $name => $script) {
+      $cleanScript = $this->cleanup($script);
+      $attachments['#attached']['html_head'][] = [
+        [
+          '#tag' => 'script',
+          '#value' => Markup::create($cleanScript),
+          '#weight' => 1,
+        ],
+        $name . '_script',
+      ];
```
Comment (rimi-itk) [inline web/profiles/custom/os2loop/modules/os2loop_analytics/src/Helper/AnalyticsHelper.php:47]: I appreciate what you're doing here, but we may run into issues with tracking codes that require more that a single `script` element (e.g. https://developers.google.com/analytics/devguides/collection/analyticsjs#alternative_async_tag). We should consider just letting any code through without any modifications.

File: web/profiles/custom/os2loop/modules/os2loop_analytics/src/Helper/AnalyticsHelper.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+   * Get config related to os2loop analytics.
+   *
+   * @return array
+   *   A list of analytics scripts.
+   */
+  private function getAnalyticsConfig(): array {
+    $scripts = [];
+    $scripts += $this->config->get('matomo_tracking_code') ? ['matomo' => $this->config->get('matomo_tracking_code')] : [];
+    $scripts += $this->config->get('google_analytics_tracking_code') ? ['google_analytics' => $this->config->get('google_analytics_tracking_code')] : [];
+    $scripts += $this->config->get('siteimprove_tracking_code') ? ['siteimprove' => $this->config->get('siteimprove_tracking_code')] : [];
```
Comment (rimi-itk) [inline web/profiles/custom/os2loop/modules/os2loop_analytics/src/Helper/AnalyticsHelper.php:61]: I think this code will be easier to read (for a human) if written as

```php
if ($this->config->get('matomo_tracking_code')) {
  $scripts['matomo'] = $this->config->get('matomo_tracking_code');
}
if ($this->config->get('google_analytics_tracking_code')) {
  …
}
…
```

You could also go for 
```php
$scripts = [
  'matomo' => $this->config->get('matomo_tracking_code'),
  'google_analytics' => $this->config->get('google_analytics_tracking_code'),
  …
];

// Remove empty values.
return array_filter($scripts);
```

File: web/profiles/custom/os2loop/modules/os2loop_analytics/src/Form/SettingsForm.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    $this->configFactory->getEditable(static::SETTINGS_NAME)
+      ->set('matomo_tracking_code', $form_state->getValue('matomo_tracking_code'))
+      ->set('google_analytics_tracking_code', $form_state->getValue('google_analytics_tracking_code'))
+      ->set('siteimprove_tracking_code', $form_state->getValue('siteimprove_tracking_code'))
+      ->save();
+
+    parent::submitForm($form, $form_state);
```
Comment (rimi-itk) [inline web/profiles/custom/os2loop/modules/os2loop_analytics/src/Form/SettingsForm.php:107]: We should clear cache here (https://api.drupal.org/api/drupal/core%21includes%21common.inc/function/drupal_flush_all_caches/8.2.x#comment-63788) to make sure the changes take effect immediately.

Note to self: All (or most) OS2Loop settings forms should clear cache ion save.

File: web/profiles/custom/os2loop/modules/os2loop_analytics/os2loop_analytics.links.menu.yml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+os2loop_analytics.settings:
```
Comment (rimi-itk) [inline web/profiles/custom/os2loop/modules/os2loop_analytics/os2loop_analytics.links.menu.yml:1]: These settings must config ignore'd.

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: web/profiles/custom/os2loop/modules/os2loop_analytics/src/Helper/AnalyticsHelper.php (hunk 1 above)
Comment (martinyde) [reply web/profiles/custom/os2loop/modules/os2loop_analytics/src/Helper/AnalyticsHelper.php]: Removed helper all together

File: web/profiles/custom/os2loop/modules/os2loop_analytics/src/Helper/AnalyticsHelper.php (hunk 2 above)
Comment (martinyde) [reply web/profiles/custom/os2loop/modules/os2loop_analytics/src/Helper/AnalyticsHelper.php]: Removed helper all together

File: web/profiles/custom/os2loop/modules/os2loop_analytics/src/Form/SettingsForm.php (hunk 3 above)
Comment (martinyde) [reply web/profiles/custom/os2loop/modules/os2loop_analytics/src/Form/SettingsForm.php:109]: Added cache clear

File: web/profiles/custom/os2loop/modules/os2loop_analytics/os2loop_analytics.links.menu.yml (hunk 4 above)
Comment (martinyde) [reply web/profiles/custom/os2loop/modules/os2loop_analytics/os2loop_analytics.links.menu.yml:1]: Ignored config

**Review by rimi-itk — APPROVED** [review]

**Pull request merged**